src/transport/CertificateAuthenticatedSessionEstablishment.cpp

```diff
@@ -0,0 +1,415 @@
+                                                                             System::PacketBufferHandle msgBuf)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    PayloadHeader payloadHeader;
+
+    payloadHeader
+        .SetMessageType(static_cast<uint8_t>(msgType)) //
+        .SetProtocolID(Protocols::kProtocol_SecureChannel);
+
+    uint16_t headerSize              = payloadHeader.EncodeSizeBytes();
+    uint16_t actualEncodedHeaderSize = 0;
+
+    VerifyOrExit(msgBuf->EnsureReservedSize(headerSize), err = CHIP_ERROR_NO_MEMORY);
+
+    msgBuf->SetStart(msgBuf->Start() - headerSize);
+    err = payloadHeader.Encode(msgBuf->Start(), msgBuf->DataLength(), &actualEncodedHeaderSize);
+    SuccessOrExit(err);
+    VerifyOrExit(headerSize == actualEncodedHeaderSize, err = CHIP_ERROR_INTERNAL);
+
+    err = mDelegate->SendPairingMessage(PacketHeader()
+                                            .SetSourceNodeId(mLocalNodeId)
+                                            .SetDestinationNodeId(mConnectionState.GetPeerNodeId())
+                                            .SetEncryptionKeyID(mConnectionState.GetLocalKeyID()),
+                                        mConnectionState.GetPeerAddress(), std::move(msgBuf));
+    SuccessOrExit(err);
+
+exit:
+    return err;
+}
+
+CHIP_ERROR CertificateAuthenticatedSessionEstablishment::EstablishSession(const Transport::PeerAddress peerAddress, NodeId myNodeId,
+                                                                          NodeId peerNodeId, uint16_t myKeyId,
+                                                                          AuthenticatedSessionEstablishmentDelegate * delegate)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    mConnectionState.SetPeerAddress(peerAddress);
+    mConnectionState.SetPeerNodeId(peerNodeId);
+
+    err = SendSigmaR1();
+    SuccessOrExit(err);
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        Clear();
+    }
+    return err;
+}
+
+CHIP_ERROR CertificateAuthenticatedSessionEstablishment::DeriveSecureSession(const uint8_t * info, size_t info_len,
+                                                                             SecureSession & session)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    VerifyOrExit(info != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(info_len > 0, err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(mPairingComplete, err = CHIP_ERROR_INCORRECT_STATE);
+
+    // ToDo: Use asymmetric keypair to create the secure session
+    // err = session.Init();
+    SuccessOrExit(err);
+
+exit:
+    return err;
+}
+
+CHIP_ERROR CertificateAuthenticatedSessionEstablishment::SendSigmaR1()
+{
+    uint16_t data_len = 0;
+
+    System::PacketBufferHandle msg_R1;
+
+    msg_R1 = System::PacketBuffer::NewWithAvailableSize(data_len);
+    VerifyOrReturnError(!msg_R1.IsNull(), CHIP_SYSTEM_ERROR_NO_MEMORY);
+
+    // ToDo: Construct SigmaR1 message in form of the following structure
```

Comment:
Please file issue, link here.  Also, I think we are using "TODO" (all caps) for todos.